# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/base/nsFrameTraversal.h
# Commit: 0f896a4a5462
# Full Hash: 0f896a4a54629904c55cc5e18aa1f6e387c9a697
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2023-12-24 09:09:23
# Regressor Bug: 779684
# File Overlap Count: 1
# Description:
#   Bug 779684 - part 3: Make `nsFrameIterator` non-refcountable r=emilio
#   
#   Nobody shares an instance of `nsFrameIterator`.  Therefore, we can make it
#   non-refcountable.  Additionally, `nsVisualIterator` is not required because
#   it overrides only 4 methods which are really simple.  So, once we merge it
# ==============================================================================

diff -r 92148b810f22 -r 0f896a4a5462 layout/base/nsFrameTraversal.h
--- a/layout/base/nsFrameTraversal.h	Sun Dec 24 06:01:19 2023 +0000
+++ b/layout/base/nsFrameTraversal.h	Sun Dec 24 06:01:19 2023 +0000
@@ -7,15 +7,13 @@
 #ifndef NSFRAMETRAVERSAL_H
 #define NSFRAMETRAVERSAL_H
 
+#include <cstdint>
 #include "mozilla/Attributes.h"
-#include "nsISupportsImpl.h"
 
 class nsIFrame;
 class nsPresContext;
 
-class nsFrameIterator {
-  NS_INLINE_DECL_REFCOUNTING(nsFrameIterator)
-
+class MOZ_STACK_CLASS nsFrameIterator final {
  public:
   void First();
   void Next();
@@ -42,17 +40,12 @@
     // "close tag" order
     PostOrder,
   };
-  static already_AddRefed<nsFrameIterator> Create(
-      nsPresContext* aPresContext, nsIFrame* aStart, Type aType, bool aVisual,
-      bool aLockInScrollView, bool aFollowOOFs, bool aSkipPopupChecks,
-      nsIFrame* aLimiter = nullptr);
+  nsFrameIterator(nsPresContext* aPresContext, nsIFrame* aStart, Type aType,
+                  bool aVisual, bool aLockInScrollView, bool aFollowOOFs,
+                  bool aSkipPopupChecks, nsIFrame* aLimiter = nullptr);
+  ~nsFrameIterator() = default;
 
  protected:
-  nsFrameIterator(nsPresContext* aPresContext, nsIFrame* aStart, Type aType,
-                  bool aLockInScrollView, bool aFollowOOFs,
-                  bool aSkipPopupChecks, nsIFrame* aLimiter);
-  virtual ~nsFrameIterator() = default;
-
   void SetCurrent(nsIFrame* aFrame) { mCurrent = aFrame; }
   nsIFrame* GetCurrent() { return mCurrent; }
   nsIFrame* GetStart() { return mStart; }
@@ -96,11 +89,11 @@
    visual order".
   */
 
-  virtual nsIFrame* GetFirstChildInner(nsIFrame* aFrame);
-  virtual nsIFrame* GetLastChildInner(nsIFrame* aFrame);
+  nsIFrame* GetFirstChildInner(nsIFrame* aFrame);
+  nsIFrame* GetLastChildInner(nsIFrame* aFrame);
 
-  virtual nsIFrame* GetNextSiblingInner(nsIFrame* aFrame);
-  virtual nsIFrame* GetPrevSiblingInner(nsIFrame* aFrame);
+  nsIFrame* GetNextSiblingInner(nsIFrame* aFrame);
+  nsIFrame* GetPrevSiblingInner(nsIFrame* aFrame);
 
   /**
    * Return the placeholder frame for aFrame if it has one, otherwise return
@@ -115,6 +108,7 @@
   const bool mLockScroll;
   const bool mFollowOOFs;
   const bool mSkipPopupChecks;
+  const bool mVisual;
   const Type mType;
 
  private: